src/main/java/org/springframework/samples/petclinic/rest/importcsv/ImportCSV.java

```diff
@@ -26,114 +26,123 @@
-                        pet.setType(ts.get(j));
-                        break;
-                    }
-                }
-            }
+    /**
+     * Custom Exception which can arise during csv string parsing, error message
+     * contains where the exception was raised
+     */
+    private class IncorrectFormatException extends Exception {
+        public IncorrectFormatException(String errorMessage) {
+            super(errorMessage);
+        }
+    }
 
-            field = "";
-            while (i < csv.length() && (csv.charAt(i) != ';' && csv.charAt(i) != '\n')) {
-                field += csv.charAt(i++);
+    /**
+     * Method which receives a string from the csv and pars it to a list of pets
+     * which is then returns. Exceptions are thrown upon the error field
+     *
+     * @param csv the csv string from the user
+     * @return pets list of pets
+     * @throws IncorrectFormatException the user input is malformatted
+     */
+    private List<Pet> ParseFields(String csv) throws IncorrectFormatException {
+        int i = 0;
+        List<Pet> pets = new LinkedList<Pet>();
+        Pet pet;
+        pet = new Pet();
+
+        String[] arrayOfCsv = csv.split(";", 0);
+        pet.setName(arrayOfCsv[0]);
+
+        try {
+            pet.setBirthDate((new SimpleDateFormat("yyyy-MM-dd")).parse(arrayOfCsv[1]));
+        } catch (ParseException e) {
+            throw new IncorrectFormatException("Date not valid");
+        }
+
+        ArrayList<PetType> petTypes = (ArrayList<PetType>) clinicService.findPetTypes();
+        for (PetType petType : petTypes) {
+            if (petType.getName().toLowerCase().equals(arrayOfCsv[2])) {
+                pet.setType(petType);
+                break;
             }
-
-            if (pet != null) {
-                String owner = field;
-                List<Owner> matchingOwners = clinicService.findAllOwners()
-                    .stream()
-                    .filter(o -> o.getLastName().equals(owner))
-                    .collect(Collectors.toList());
-
-                if (matchingOwners.size() == 0) {
-                    HttpHeaders headers = new HttpHeaders();
-                    headers.add("errors", "Owner not found");
-                    return new ResponseEntity<List<Pet>>(headers, HttpStatus.BAD_REQUEST);
+        }
+
+        String owner = arrayOfCsv[3];
+        List<Owner> matchingOwners = clinicService.findAllOwners()
+            .stream()
+            .filter(o -> o.getLastName().equals(owner))
+            .collect(Collectors.toList());
+
+        if (matchingOwners.size() == 0) {
+            throw new IncorrectFormatException("Owner not found");
+        }
+        if (matchingOwners.size() > 1) {
+            throw new IncorrectFormatException("Owner not unique");
+        }
+        pet.setOwner(matchingOwners.iterator().next());
+
+        if (arrayOfCsv[4].toLowerCase().equals("add")) {
+            clinicService.savePet(pet);
+        } else {
```

Comment:
if the operation was something else (e.g., `add2`), it would also branch to the else clause, which may not be desired.